src/aalwines/model/RoutingTable.cpp

```diff
@@ -32,12 +32,125 @@
+        if (_ops.empty()) {
+            _ops.push_back(action);
+            return;
+        }
+        switch (action._op) {
+            case PUSH:
+                if (_ops.back()._op == POP) {
+                    _ops.pop_back();
+                    add_action(action_t{SWAP, action._op_label});
+                    return;
+                }
+                break;
+            case SWAP: // TODO: case SWAP and case POP doesn't seem happen, so do we want to keep it?
+                if (_ops.back()._op == SWAP || _ops.back()._op == PUSH) {
+                    _ops.back()._op_label = action._op_label;
+                    return;
+                }
+                break;
+            case POP:
+                if (_ops.back()._op == SWAP) {
+                    _ops.pop_back();
+                    add_action(action_t{POP, label_t{}});
+                    return;
+                }
+                break;
+            default:
+                break;
+        }
+        _ops.push_back(action);
+    }
+    void RoutingTable::add_failover_entries(const Interface* failed_inf, Interface* backup_inf, label_t failover_label) {
+        for (auto& e : _entries) {
+            std::vector<forward_t> new_rules;
+            for (const auto& f : e._rules) {
+                if (f._via == failed_inf) {
+                    new_rules.emplace_back(f._type, f._ops, backup_inf, f._weight + 1);
+                    new_rules.back().add_action(action_t{PUSH, failover_label});
+                }
+            }
+            e._rules.insert(e._rules.end(), new_rules.begin(), new_rules.end());
+        }
+    }
+    void RoutingTable::entry_t::add_to_outgoing(const Interface *outgoing, action_t action) {
+        for (auto&& f : _rules) {
+            if (f._via == outgoing) {
+                f.add_action(action);
+            }
+        }
+    }
+    void RoutingTable::add_to_outgoing(const Interface* outgoing, action_t action) {
+        for (auto& e : _entries) {
+            e.add_to_outgoing(outgoing, action);
+        }
+    }
+
+    void RoutingTable::simple_merge(const RoutingTable& other) {
+        assert(std::is_sorted(other._entries.begin(), other._entries.end()));
+        assert(std::is_sorted(_entries.begin(), _entries.end()));
+        auto iit = _entries.begin();
+        for (const auto & e : other._entries) {
+            while (iit != std::end(_entries) && (*iit) < e) ++iit;
+            if (iit == std::end(_entries)) {
+                iit = _entries.insert(iit, e);
+            } else if ((*iit) == e) {
+                if (e._rules.size() == 1 && iit->_rules.size() == 1 &&
+                    e._rules[0]._type == iit->_rules[0]._type && iit->_rules[0]._type != MPLS)
+                    continue;
+                bool legal_merge = true;
+                for (auto&& rule: e._rules) { // TODO: Consider sorted rules for faster merge.
```

Comment:
Changed to const auto &, which is the correct way here. 

auto&& is a forwarding reference, and I think that in this case, it will become an l-value reference, so the code would be fine, but I am dangerously unsure.